File: 2021/day11/solution.py
from typing import List

class Octopus:

    def __init__(self, val: int, flashed: bool = False) -> None:
        self.val = val
        self.flashed = flashed

class Grid:

    def __init__(self, nums: List[List[int]]):
        self.octo: List[List[Octopus]] = []
        self.threshold = 9

        for r in range(len(nums)):
            row = []
            for c in range(len(nums[0])):
                row.append(Octopus(val=nums[r][c]))
            self.octo.append(row)

    def flash(self, r: int, c: int) -> None:
        if r < 0 or c < 0 or r >= len(self.octo) or c >= len(self.octo[0]):
            return
        elif self.octo[r][c].flashed:
            return
        else:
            self.octo[r][c].val += 1
            if self.octo[r][c].val > self.threshold:
                self.octo[r][c].val = 0
                self.octo[r][c].flashed = True
                adj = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1],  [1, 0],  [1, 1]]
                for r_offset, c_offset in adj:
                    self.flash(r+r_offset, c+c_offset)
    
    def reset_flashes(self) -> int:
        flashed = 0
        for r in range(len(self.octo)):
            for c in range(len(self.octo[0])):
                if self.octo[r][c].flashed:
                    flashed += 1
                    self.octo[r][c].flashed = False
        return flashed
    
    def increment(self) -> List[List[int]]:
        to_flash = []
        for r in range(len(self.octo)):
            for c in range(len(self.octo[0])):
                self.octo[r][c].val += 1
                if self.octo[r][c].val > self.threshold:
                    to_flash.append([r,c])
        return to_flash


    def cycle(self) -> int:
        to_flash = self.increment()
        for r,c in to_flash:
            self.flash(r,c)
        return self.reset_flashes()



def get_grid() -> Grid:
    with open('input.txt') as f:
        lines = f.readlines()
        data = []
        for l in lines:
            row = []
            numstrs = l.strip()
            for s in numstrs:
                row.append(int(s.strip()))
            data.append(row)
        return Grid(nums=data)

def get_flash_count(g: Grid, cycles: int):
    total = 0
    for _ in range(cycles):
        total += g.cycle()
    return total

def get_simultaneous_count(g: Grid) -> int:
    cycles = 0
    total_cells = len(g.octo) * len(g.octo[0])
    while True:
        cycles += 1
        flashes = g.cycle()
        if flashes == total_cells:
            return cycles


p1 = get_flash_count(get_grid(), 100)
print(p1)

p2 = get_simultaneous_count(get_grid())
print(p2)

File: 2021/day11/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <queue>
#include <set>

using namespace std;

const int FLASH_THRESHOLD = 10;

vector<vector<int>> getEnergyLevels(const string& filePath) {
    ifstream inputFile(filePath);
    vector<string> lines;
    if (!inputFile) {
        cerr << "Failed to open input file." << endl;
        return {};
    }
    string line;
    while (getline(inputFile, line)) {
        lines.push_back(line);
    }
    vector<vector<int>> heightMap = {};
    for (const auto& l : lines) {
        vector<int> row = {};
        for (auto c : l) {
            // Zero must be subtracted to account for ASCII conversion
            int val = c - '0';
            row.push_back(val);
        }
        heightMap.push_back(row);
    }
    return heightMap;
}

void increase(vector<vector<int>>& energyLevels, set<pair<int,int>>& flashed, int r, int c) {
    if (r < 0 || c < 0 || r >= energyLevels.size() || c >= energyLevels[0].size() || flashed.count(make_pair(r,c))) {
        return;
    }
    energyLevels[r][c] += 1;
}

int flash(vector<vector<int>>& energyLevels) {
    queue<pair<int,int>> toFlash;
    set<pair<int,int>> flashed;

    for (auto r = 0; r < energyLevels.size(); r++) {
        for (auto c = 0; c < energyLevels[0].size(); c++) {
           energyLevels[r][c] += 1;
            if (energyLevels[r][c] >= FLASH_THRESHOLD) {
                toFlash.emplace(r,c);
            }
        }
    }
    while (!toFlash.empty()) {
        // process the current queue
        while (!toFlash.empty()) {
            pair<int, int> p = toFlash.front();
            toFlash.pop();
            int r = p.first;
            int c = p.second;
            increase(energyLevels, flashed, r-1, c-1);
            increase(energyLevels, flashed, r-1, c);
            increase(energyLevels, flashed, r-1, c+1);
            increase(energyLevels, flashed, r, c+1);
            increase(energyLevels, flashed, r+1, c+1);
            increase(energyLevels, flashed, r+1, c);
            increase(energyLevels, flashed, r+1, c-1);
            increase(energyLevels, flashed, r, c-1);
            energyLevels[r][c] = 0;
            flashed.insert(p);
        }
        // add to queue to determine if there are additional flashes that must happen
        for (auto r = 0; r < energyLevels.size(); r++) {
            for (auto c = 0; c < energyLevels[0].size(); c++) {
                auto p = make_pair(r,c);
                if (energyLevels[r][c] >= FLASH_THRESHOLD && flashed.count(p) == 0) {
                    toFlash.emplace(p);
                }
            }
        }
    }
    return flashed.size();
}

int getTotalFlashesAfterNSteps(vector<vector<int>> energyLevels, int n) {
    int flashes = 0;
    for (auto i = 0; i < n; i++) {
        flashes += flash(energyLevels);
    }
    return flashes;
}

int getFirstSynchronizedFlash(vector<vector<int>> energyLevels) {
    int flashed = 0;
    int cycles = 0;
    while (flashed != 100) {
        flashed = flash(energyLevels);
        cycles += 1;
    }
    return cycles;
}

int main() {
    auto energyLevels = getEnergyLevels("input.txt");

    cout << "Part 1: " + to_string(getTotalFlashesAfterNSteps(energyLevels, 100)) << endl;
    cout << "Part 2: " + to_string(getFirstSynchronizedFlash(energyLevels)) << endl;
}